File: algorithm/oc/flatten-nested-list-iterator.m
// 341. Flatten Nested List Iterator
// Difficulty: Medium

// Given a nested list of integers, implement an iterator to flatten it.

// Each element is either an integer, or a list -- whose elements may also be integers or other lists.

// Example 1:
// Given the list [[1,1],2,[1,1]],

// By calling next repeatedly until hasNext returns false, the order of elements returned by next should be: [1,1,2,1,1].

// Example 2:
// Given the list [1,[4,[6]]],

// By calling next repeatedly until hasNext returns false, the order of elements returned by next should be: [1,4,6].

// Time:  O(n), n is the number of the integers.
// Space: O(h), h is the depth of the nested lists.

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */


#import <Foundation/Foundation.h>

#pragma mark Stack

@interface Stack : NSObject

- (instancetype)init;
- (id)pop;
- (id)top;
- (void)push:(id)element;
- (BOOL)isEmpty;
-(NSInteger)count;

@end

#pragma mark NestedInteger

@interface NestedInteger : NSObject

-(int)getInteger;
-(BOOL)isInteger;
-(NSArray*)getList;

@end

#pragma mark Solution

@interface NestedIterator : NSObject
@end

@implementation NestedIterator

Stack* _depth;

-(instancetype)initWithNestedInteger:(NSArray*)nestedList {
  self = [super init];
  if (self) {
    [_depth push:[nestedList objectEnumerator]];
  }
  return self;
}

-(int)next {
  return [[[_depth top] nextObject] getInteger];
}

-(BOOL)hasNext {
  while (![_depth isEmpty]) {
    NestedInteger* object = [[_depth top] nextObject];
    if (!object) {
      [_depth pop];
    } else if ([object isInteger]) {
      return YES;
    } else {
      [_depth push:[[object getList] objectEnumerator]];
    }
  }
  return NO;
}

@end

File: data-structure-and-algorithm/typescript/leetcode/src/flatten-nested-list-iterator.ts
/*
You are given a nested list of integers nestedList. Each element is either an integer or a list whose elements may also be integers or other lists. Implement an iterator to flatten it.

Implement the NestedIterator class:

NestedIterator(List<NestedInteger> nestedList) Initializes the iterator with the nested list nestedList.
int next() Returns the next integer in the nested list.
boolean hasNext() Returns true if there are still some integers in the nested list and false otherwise.
Your code will be tested with the following pseudocode:

initialize iterator with nestedList
res = []
while iterator.hasNext()
    append iterator.next() to the end of res
return res
If res matches the expected flattened list, then your code will be judged as correct.



Example 1:

Input: nestedList = [[1,1],2,[1,1]]
Output: [1,1,2,1,1]
Explanation: By calling next repeatedly until hasNext returns false, the order of elements returned by next should be: [1,1,2,1,1].
Example 2:

Input: nestedList = [1,[4,[6]]]
Output: [1,4,6]
Explanation: By calling next repeatedly until hasNext returns false, the order of elements returned by next should be: [1,4,6].


Constraints:

1 <= nestedList.length <= 500
The values of the integers in the nested list is in the range [-106, 106].
*/

/**
 * Your NestedIterator will be called like this:
 * var i = new NestedIterator(nestedList), a = [];
 * while (i.hasNext()) a.push(i.next());
 */

class NestedIterator {
  constructor(list) {
    this.stack = [];
    list.reverse().forEach(entry => this.stack.push(entry));
  }

  hasNext() {
    while (this.stack.length) {
      const top = this.stack[this.stack.length - 1];
      if (top.isInteger()) {
        return true;
      }
      this.stack.pop();

      top
        .getList()
        .reverse()
        .forEach(entry => this.stack.push(entry));
    }

    return false;
  }

  next() {
    return this.stack.pop().getInteger();
  }
}
